File: StudyNotes/GeekBang.org/视频课/算法面试通关40讲/08 | 面试题：判断括号字符串是否有效/0020/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        paren_map = {')': '(', ']': '[', '}': '{'}
        for c in s:
            if c not in paren_map:
                stack.append(c)
            elif not stack or paren_map[c] != stack.pop():
                return False
        return not stack

File: StudyNotes/GeekBang.org/视频课/算法面试通关40讲/08 | 面试题：判断括号字符串是否有效/0020/20.java
class Solution {
    public boolean isValid(String s) {
        char[] chars = s.toCharArray();
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> paren_map = new HashMap<>();
        paren_map.put(')', '(');
        paren_map.put(']', '[');
        paren_map.put('}', '{');
        for (int i = 0; i < chars.length; i++) {
            Character c = chars[i];
            if (!paren_map.containsKey(c)) {
                stack.push(c);
            } else if (stack.isEmpty() || !paren_map.get(c).equals(stack.pop())) {
                return false;
            }
        }
        return stack.isEmpty();
    }
}